validator/client/wait_for_activation.go

```diff
@@ -40,99 +41,132 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
 	ctx, span := trace.StartSpan(ctx, "validator.WaitForActivation")
 	defer span.End()
+
+	// Step 1: Fetch validating public keys.
 	validatingKeys, err := v.km.FetchValidatingPublicKeys(ctx)
 	if err != nil {
 		return errors.Wrap(err, msgCouldNotFetchKeys)
 	}
-	// if there are no validating keys, wait for some
+
+	// Step 2: If no keys, wait for accounts change or context cancellation.
 	if len(validatingKeys) == 0 {
 		log.Warn(msgNoKeysFetched)
-		select {
-		case <-ctx.Done():
-			log.Debug("Context closed, exiting fetching validating keys")
-			return ctx.Err()
-		case <-accountsChangedChan:
-			// if the accounts changed try it again
-			return v.internalWaitForActivation(ctx, accountsChangedChan)
-		}
+		return v.waitForAccountsChange(ctx, accountsChangedChan)
+	}
+
+	// Step 3: update validator statuses in cache.
+	if err := v.updateValidatorStatusCache(ctx, validatingKeys); err != nil {
+		return v.handleReconnection(ctx, span, err, "Connection broken while waiting for activation. Reconnecting...", accountsChangedChan)
 	}
 
-	stream, err := v.validatorClient.WaitForActivation(ctx, &ethpb.ValidatorActivationRequest{
-		PublicKeys: bytesutil.FromBytes48Array(validatingKeys),
-	})
+	// Step 4: Fetch validator count.
+	valCount, err := v.getValidatorCount(ctx)
 	if err != nil {
-		tracing.AnnotateError(span, err)
-		attempts := streamAttempts(ctx)
-		log.WithError(err).WithField("attempts", attempts).
-			Error("Stream broken while waiting for activation. Reconnecting...")
-		// Reconnection attempt backoff, up to 60s.
-		time.Sleep(time.Second * time.Duration(math.Min(uint64(attempts), 60)))
-		return v.internalWaitForActivation(incrementRetries(ctx), accountsChangedChan)
+		return err
+	}
+
+	// Step 5: Check and log validator statuses.
+	someAreActive := v.checkAndLogValidatorStatus(valCount)
+	if !someAreActive {
+		// Step 6: If no active validators, wait for accounts change, context cancellation, or next epoch.
+		return v.waitForActivationRetry(ctx, span, accountsChangedChan)
 	}
+	return nil
+}
+
+func (v *validator) getValidatorCount(ctx context.Context) (int64, error) {
+	// TODO: revisit https://github.com/prysmaticlabs/prysm/pull/12471#issuecomment-1568320970 to review if ValidatorCount api can be removed.
+	// "-1" indicates that validator count endpoint is not supported by the beacon node.
+	var valCount int64 = -1
+	valCounts, err := v.prysmChainClient.ValidatorCount(ctx, "head", []validator2.Status{validator2.Active})
+	if err != nil && !errors.Is(err, iface.ErrNotSupported) {
+		return -1, errors.Wrap(err, "could not get active validator count")
+	}
+	if len(valCounts) > 0 {
+		valCount = int64(valCounts[0].Count)
+	}
+	return valCount, nil
+}
+
+func (v *validator) handleReconnection(ctx context.Context, span octrace.Span, err error, message string, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
+	tracing.AnnotateError(span, err)
+	attempts := streamAttempts(ctx)
```

Comment:
should i rename this streamAttempts variable?

Reply:
The code block is  small enough that I don't think it matters too much. It should be very clear what it is

Reply:
i guess it's not really a stream